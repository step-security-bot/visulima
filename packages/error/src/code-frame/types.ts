export type CodeFrameLocation = {
    column?: number;
    line: number;
};

export type CodeFrameNodeLocation = {
    end?: CodeFrameLocation;
    start: CodeFrameLocation;
};

export type ColorizeMethod = (value: string) => string;

export type CodeFrameOptions = {
    color?: Partial<{
        gutter: ColorizeMethod;
        marker: ColorizeMethod;
        message: ColorizeMethod;
    }>;
    linesAbove?: number;
    linesBelow?: number;
    message?: string;
    showGutter?: boolean;
    showLineNumbers?: boolean;
};
