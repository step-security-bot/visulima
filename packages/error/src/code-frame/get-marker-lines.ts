/**
 * This is a copy of the codeFrame function from Babel
 * @see https://github.com/babel/babel/blob/85e649203b61b7c908eb04c05511a0d35f893e8e/packages/babel-code-frame/src/index.ts#L68-L143
 *
 * MIT License
 *
 * Copyright (c) 2014-present Sebastian McKenzie and other contributors
 */

/**
 * Extract what lines should be marked and highlighted.
 */
import type { CodeFrameLocation, CodeFrameNodeLocation } from "../types";

type MarkerLines = Record<number, true | [number | undefined, number | undefined]>;

const getMarkerLines = (
    loc: CodeFrameNodeLocation,
    source: Array<string>,
    linesAbove: number,
    linesBelow: number,
): {
    start: number;
    end: number;
    markerLines: MarkerLines;
} => {
    const startLoc: CodeFrameLocation = {
        column: 0,
        // @ts-expect-error Can be overwritten
        line: -1,
        ...loc.start,
    };
    const endLoc: CodeFrameLocation = {
        ...startLoc,
        ...loc.end,
    };
    const startLine = startLoc.line;
    const startColumn = startLoc.column;
    const endLine = endLoc.line;
    const endColumn = endLoc.column;

    let start = Math.max(startLine - (linesAbove + 1), 0);
    let end = Math.min(source.length, endLine + linesBelow);

    if (startLine === -1) {
        start = 0;
    }

    if (endLine === -1) {
        end = source.length;
    }

    const lineDiff = endLine - startLine;
    const markerLines: MarkerLines = {};

    if (lineDiff) {
        for (let i = 0; i <= lineDiff; i++) {
            const lineNumber = i + startLine;

            if (!startColumn) {
                markerLines[lineNumber] = true;
            } else if (i === 0) {
                const sourceLength = source[lineNumber - 1]?.length;

                markerLines[lineNumber] = [startColumn, (sourceLength || 0) - startColumn + 1];
            } else if (i === lineDiff) {
                markerLines[lineNumber] = [0, endColumn];
            } else {
                const sourceLength = source[lineNumber - i]?.length;

                markerLines[lineNumber] = [0, sourceLength];
            }
        }
    } else {
        if (startColumn === endColumn) {
            if (startColumn) {
                markerLines[startLine] = [startColumn, 0];
            } else {
                markerLines[startLine] = true;
            }
        } else {
            markerLines[startLine] = [startColumn, (endColumn || 0) - (startColumn || 0)];
        }
    }

    return { start, end, markerLines };
};

export default getMarkerLines;
