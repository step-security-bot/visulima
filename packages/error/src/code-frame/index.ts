/**
 * This is a modified copy of the codeFrameColumns function from Babel
 * @see https://github.com/babel/babel/blob/85e649203b61b7c908eb04c05511a0d35f893e8e/packages/babel-code-frame/src/index.ts#L145-L217
 *
 * MIT License
 *
 * Copyright (c) 2014-present Sebastian McKenzie and other contributors
 */

import process from "../util/process";
import normalizeLF from "../util/normalize-lf";
import type { CodeFrameNodeLocation, CodeFrameOptions, ColorizeMethod } from "../types";
import getMarkerLines from "./get-marker-lines";

const POINTER = process.platform === "win32" && !process.env?.["WT_SESSION"] ? ">" : "‚ùØ";

/** Generate a code frame from string and an error location */
export const codeFrame = (
    source: string,
    loc: CodeFrameNodeLocation,
    options: Partial<CodeFrameOptions> = {},
    // eslint-disable-next-line sonarjs/cognitive-complexity
): string => {
    // grab 2 lines before, and 3 lines after focused line
    const config = {
        linesAbove: 2,
        linesBelow: 3,
        showLineNumbers: true,
        showGutter: true,
        color: {
            gutter: (value: string) => value,
            marker: (value: string) => value,
            message: (value: string) => value,
            ...options?.color,
        },
        ...options,
    };

    const hasColumns = loc.start && typeof loc.start.column === "number";

    const lines = normalizeLF(source)
        .split("\n")
        .map((ln) => ln.replaceAll("\t", "    "));

    const { start, end, markerLines } = getMarkerLines(loc, lines, config.linesAbove, config.linesBelow);

    const numberMaxWidth = String(end).length;

    const {
        message: colorizeMessage,
        marker: colorizeMarker,
        gutter: colorizeGutter,
    } = config.color as {
        gutter: ColorizeMethod;
        marker: ColorizeMethod;
        message: ColorizeMethod;
    };

    let frame = lines
        .slice(start, end)
        .map((line, index) => {
            const number = start + 1 + index;
            const hasMarker = markerLines[number];
            const paddedNumber = ` ${number}`.slice(-numberMaxWidth);
            const lastMarkerLine = !markerLines[number + 1];

            const gutter = ` ${paddedNumber}${config.showGutter ? ` |` : ""}`;

            if (hasMarker) {
                let markerLine = "";

                if (Array.isArray(hasMarker)) {
                    const markerSpacing = line.replace(/[^\t]/g, " ").slice(0, Math.max((hasMarker[0] as number) - 1, 0));
                    const numberOfMarkers = hasMarker[1] || 1;

                    markerLine = ["\n ", colorizeGutter(gutter.replace(/\d/g, " ")), " ", markerSpacing, colorizeMarker("^").repeat(numberOfMarkers)].join("");

                    if (lastMarkerLine && config?.message) {
                        markerLine += " " + colorizeMessage(config.message);
                    }
                }

                return [colorizeMarker(POINTER), colorizeGutter(gutter), line.length > 0 ? ` ${line}` : "", markerLine].join("");
            }

            return ` ${colorizeGutter(gutter)}${line.length > 0 ? ` ${line}` : ""}`;
        })
        .join("\n");

    if (config?.message && !hasColumns) {
        frame = `${" ".repeat(numberMaxWidth + 1)}${config.message}\n${frame}`;
    }

    return frame;
};

export type { CodeFrameOptions, CodeFrameNodeLocation, ColorizeMethod } from "../types";
