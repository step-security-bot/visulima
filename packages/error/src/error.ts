export type { ErrorLocation, ErrorProperties, ErrorWithMetadata, Trace } from "./types";
export { isVisulimaError, VisulimaError } from "./visulima-error";
