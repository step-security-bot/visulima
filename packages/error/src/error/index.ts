export { default as getErrorCauses } from "./get-error-causes";
export type { ErrorHint, ErrorLocation, ErrorProperties } from "./types";
export { isVisulimaError, VisulimaError } from "./visulima-error";
