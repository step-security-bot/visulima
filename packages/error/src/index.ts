export type { CodeFrameLocation, CodeFrameNodeLocation, CodeFrameOptions, ColorizeMethod } from "./code-frame";
export { CODE_FRAME_POINTER, codeFrame } from "./code-frame";
export type { ErrorHint, ErrorLocation, ErrorProperties } from "./error";
export { getErrorCauses, isVisulimaError, VisulimaError } from "./error";
export type { TraceMap } from "./sourcemap";
export { generatedPositionFor, loadSourceMap, originalPositionFor, sourceContentFor, traceSegment } from "./sourcemap";
export type { Trace, TraceType } from "./stacktrace";
export { parseStacktrace } from "./stacktrace";
