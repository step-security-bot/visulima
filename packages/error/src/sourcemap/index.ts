export type { TraceMap } from "./sourcemap";
export { generatedPositionFor, loadSourceMap, originalPositionFor, sourceContentFor, traceSegment } from "./sourcemap";
