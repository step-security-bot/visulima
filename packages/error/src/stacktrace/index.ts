export { default as parseStacktrace } from "./parse-stacktrace";
export type { Trace, TraceType } from "./types";
