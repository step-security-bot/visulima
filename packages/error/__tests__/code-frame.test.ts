/**
 * This is a copy of the codeFrameColumns tests from Babel
 * @see https://github.com/babel/babel/blob/85e649203b61b7c908eb04c05511a0d35f893e8e/packages/babel-code-frame/test/index.js#L316-L565
 * @see https://github.com/babel/babel/blob/85e649203b61b7c908eb04c05511a0d35f893e8e/packages/babel-code-frame/test/index.js#L71-L139
 *
 * MIT License
 *
 * Copyright (c) 2014-present Sebastian McKenzie and other contributors
 */

import chalk from "chalk";
import { describe, expect, it, vi } from "vitest";

import process from "../src/util/process";
import { codeFrame } from "../src/code-frame";

const POINTER = process.platform === "win32" && !process.env?.["WT_SESSION"] ? ">" : "❯";

vi.mock("./utils", () => {
    return {
        normalizeLF: (string_: string) => string_,
    };
});

describe("code-frame", () => {
    it("should return an empty string if line or column is undefined", () => {
        const source = "const x = 10;\nconst error = x.y;\n";
        const loc = { column: undefined, line: undefined };
        const result = codeFrame(source, { start: loc });

        expect(result).toBe("");
    });

    it("should return the correct code frame for a given location", () => {
        const source = "const x = 10;\nconst error = x.y;\n";
        const loc = { column: 16, line: 2 };
        const result = codeFrame(source, { start: loc });

        expect(result).toBe(`  1 | const x = 10;
${POINTER} 2 | const error = x.y;
    |                ^
  3 |`);
    });

    it("should handle long result", () => {
        const source =
            "function getUser () {\n    const error = new Error('Unable to find user', {\n        cause: new Error('foo')\n    })\n    error.help = [\n    'We tried looking for using inside the \"users\" table',\n    'The search was performed using the where (email = user.email) and (is_active = true)'\n   ]\n\n    throw error\n}";
        const loc = { column: 5, line: 8 };
        const result = codeFrame(source, { start: loc });

        expect(result).toBe(`   6 |     'We tried looking for using inside the "users" table',
   7 |     'The search was performed using the where (email = user.email) and (is_active = true)'
${POINTER}  8 |    ]
     |     ^
   9 |
  10 |     throw error
  11 | }`);
    });

    it("should be possible to change the visible lines", () => {
        const source =
            "function getUser () {\n    const error = new Error('Unable to find user', {\n        cause: new Error('foo')\n    })\n    error.help = [\n    'We tried looking for using inside the \"users\" table',\n    'The search was performed using the where (email = user.email) and (is_active = true)'\n   ]\n\n    throw error\n}";
        const loc = { column: 5, line: 8 };
        const result = codeFrame(
            source,
            { start: loc },
            {
                linesAbove: 1,
                linesBelow: 2,
            },
        );

        expect(result).toBe(`   7 |     'The search was performed using the where (email = user.email) and (is_active = true)'
${POINTER}  8 |    ]
     |     ^
   9 |
  10 |     throw error`);
    });

    it("should colorize the error line", () => {
        const source =
            "function getUser () {\n    const error = new Error('Unable to find user', {\n        cause: new Error('foo')\n    })\n    error.help = [\n    'We tried looking for using inside the \"users\" table',\n    'The search was performed using the where (email = user.email) and (is_active = true)'\n   ]\n\n    throw error\n}";
        const loc = { column: 5, line: 8 };
        const result = codeFrame(
            source,
            { start: loc },
            {
                color: {
                    gutter: chalk.grey,
                    marker: chalk.red.bold,
                    message: chalk.red.bold,
                },
            },
        );

        expect(result).toMatchSnapshot();
    });

    it("should show no lines above", function () {
        const rawLines = ["class Foo {", "  constructor() {", "    console.log(arguments);", "  }", "};"].join("\n");

        expect(codeFrame(rawLines, { start: { line: 2 } }, { linesAbove: 0 })).toEqual(
            [`${POINTER} 2 |   constructor() {`, "  3 |     console.log(arguments);", "  4 |   }", "  5 | };"].join("\n"),
        );
    });

    it("should show no lines below", function () {
        const rawLines = ["class Foo {", "  constructor() {", "    console.log(arguments);", "  }", "};"].join("\n");

        expect(codeFrame(rawLines, { start: { line: 2 } }, { linesBelow: 0 })).toEqual(["  1 | class Foo {", `${POINTER} 2 |   constructor() {`].join("\n"));
    });

    it("should show single line", function () {
        const rawLines = ["class Foo {", "  constructor() {", "    console.log(arguments);", "  }", "};"].join("\n");

        expect(codeFrame(rawLines, { start: { line: 2 } }, { linesAbove: 0, linesBelow: 0 })).toEqual([`${POINTER} 2 |   constructor() {`].join("\n"));
    });

    it("should mark multiple columns across lines", function () {
        const rawLines = ["class Foo {", "  constructor() {", "  }", "};"].join("\n");

        expect(
            codeFrame(rawLines, {
                start: { line: 2, column: 17 },
                end: { line: 3, column: 3 },
            }),
        ).toEqual(
            ["  1 | class Foo {", `${POINTER} 2 |   constructor() {`, "    |                 ^", `${POINTER} 3 |   }`, "    | ^^^", "  4 | };"].join("\n"),
        );
    });

    it("should mark multiple columns across multiple lines", function () {
        const rawLines = ["class Foo {", "  constructor() {", "    console.log(arguments);", "  }", "};"].join("\n");

        expect(
            codeFrame(rawLines, {
                start: { line: 2, column: 17 },
                end: { line: 4, column: 3 },
            }),
        ).toEqual(
            [
                "  1 | class Foo {",
                `${POINTER} 2 |   constructor() {`,
                "    |                 ^",
                `${POINTER} 3 |     console.log(arguments);`,
                "    | ^^^^^^^^^^^^^^^^^^^^^^^^^^^",
                `${POINTER} 4 |   }`,
                "    | ^^^",
                "  5 | };",
            ].join("\n"),
        );
    });

    it("should mark across multiple lines without columns", function () {
        const rawLines = ["class Foo {", "  constructor() {", "    console.log(arguments);", "  }", "};"].join("\n");

        expect(codeFrame(rawLines, { start: { line: 2 }, end: { line: 4 } })).toEqual(
            ["  1 | class Foo {", `${POINTER} 2 |   constructor() {`, `${POINTER} 3 |     console.log(arguments);`, `${POINTER} 4 |   }`, "  5 | };"].join(
                "\n",
            ),
        );
    });

    it("should show message", function () {
        const rawLines = ["class Foo {", "  constructor()", "};"].join("\n");

        expect(
            codeFrame(
                rawLines,
                { start: { line: 2, column: 16 } },
                {
                    message: "Missing {",
                },
            ),
        ).toEqual(["  1 | class Foo {", `${POINTER} 2 |   constructor()`, "    |                ^ Missing {", "  3 | };"].join("\n"));
    });

    it("should show message without column", function () {
        const rawLines = ["class Foo {", "  constructor()", "};"].join("\n");

        expect(
            codeFrame(
                rawLines,
                { start: { line: 2 } },
                {
                    message: "Missing {",
                },
            ),
        ).toEqual(["  Missing {", "  1 | class Foo {", `${POINTER} 2 |   constructor()`, "  3 | };"].join("\n"));
    });

    it("should show message with multiple lines", function () {
        const rawLines = ["class Foo {", "  constructor() {", "    console.log(arguments);", "  }", "};"].join("\n");

        expect(
            codeFrame(
                rawLines,
                {
                    start: { line: 2, column: 17 },
                    end: { line: 4, column: 3 },
                },
                {
                    message: "something about the constructor body",
                },
            ),
        ).toEqual(
            [
                "  1 | class Foo {",
                `${POINTER} 2 |   constructor() {`,
                "    |                 ^",
                `${POINTER} 3 |     console.log(arguments);`,
                "    | ^^^^^^^^^^^^^^^^^^^^^^^^^^^",
                `${POINTER} 4 |   }`,
                "    | ^^^ something about the constructor body",
                "  5 | };",
            ].join("\n"),
        );
    });

    it("should show message with multiple lines without columns", function () {
        const rawLines = ["class Foo {", "  constructor() {", "    console.log(arguments);", "  }", "};"].join("\n");

        expect(
            codeFrame(
                rawLines,
                { start: { line: 2 }, end: { line: 4 } },
                {
                    message: "something about the constructor body",
                },
            ),
        ).toEqual(
            [
                "  something about the constructor body",
                "  1 | class Foo {",
                `${POINTER} 2 |   constructor() {`,
                `${POINTER} 3 |     console.log(arguments);`,
                `${POINTER} 4 |   }`,
                "  5 | };",
            ].join("\n"),
        );
    });

    it("should maximum context lines and padding", function () {
        const rawLines = [
            "/**",
            " * Sums two numbers.",
            " *",
            " * @param a Number",
            " * @param b Number",
            " * @returns Number",
            " */",
            "",
            "function sum(a, b) {",
            "  return a + b",
            "}",
        ].join("\n");

        expect(codeFrame(rawLines, { start: { line: 7, column: 2 } })).toEqual(
            [
                "   5 |  * @param b Number",
                "   6 |  * @returns Number",
                `${POINTER}  7 |  */`,
                "     |  ^",
                "   8 |",
                "   9 | function sum(a, b) {",
                "  10 |   return a + b",
            ].join("\n"),
        );
    });

    it("should no unnecessary padding due to one-off errors", function () {
        const rawLines = [
            "/**",
            " * Sums two numbers.",
            " *",
            " * @param a Number",
            " * @param b Number",
            " * @returns Number",
            " */",
            "",
            "function sum(a, b) {",
            "  return a + b",
            "}",
        ].join("\n");

        expect(codeFrame(rawLines, { start: { line: 6, column: 2 } })).toEqual(
            [
                "  4 |  * @param a Number",
                "  5 |  * @param b Number",
                `${POINTER} 6 |  * @returns Number`,
                "    |  ^",
                "  7 |  */",
                "  8 |",
                "  9 | function sum(a, b) {",
            ].join("\n"),
        );
    });

    it("should handle tabs correctly by replacing them with 4 spaces", () => {
        const source = "const x = 10;\n\tconst error = x.y;\n";
        const loc = { column: 20, line: 2 };
        const result = codeFrame(source, { start: loc });

        expect(result).toBe(`  1 | const x = 10;
${POINTER} 2 |     const error = x.y;
    |                    ^
  3 |`);
    });

    it("should handle tabs", () => {
        const rawLines = ["\tclass Foo {", "\t  \t\t    constructor\t(\t)", "\t};"].join("\n");

        expect(codeFrame(rawLines, { start: { line: 2, column: 25 } })).toEqual(
            ["  1 |     class Foo {", `${POINTER} 2 |                   constructor    (    )`, "    |                         ^", "  3 |     };"].join("\n"),
        );
    });

    it("should correctly marks lines containing tab characters", () => {
        expect(
            codeFrame(" * @name        Foo#a", {
                start: { line: 1, column: 17 },
                end: { line: 1, column: 19 },
            }),
        ).toEqual(`${POINTER} 1 |  * @name        Foo#a
    |                 ^^`);

        expect(
            codeFrame(" * @name\t\tFoo#a", {
                start: { line: 1, column: 17 },
                end: { line: 1, column: 19 },
            }),
        ).toEqual(`${POINTER} 1 |  * @name        Foo#a
    |                 ^^`);
    });
});
