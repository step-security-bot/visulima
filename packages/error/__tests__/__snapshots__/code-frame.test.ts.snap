// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`code-frame > should colorize the error line 1`] = `
" [90m  6 |[39m     'We tried looking for using inside the \\"users\\" table',
 [90m  7 |[39m     'The search was performed using the where (email = user.email) and (is_active = true)'
[31m[1m‚ùØ[22m[39m[90m  8 |[39m    ]
 [90m    |[39m     [31m[1m^[22m[39m
 [90m  9 |[39m
 [90m 10 |[39m     throw error
 [90m 11 |[39m }"
`;
